language/changes/5-named-addresses.md

```diff
@@ -0,0 +1,212 @@
+syntax reduces the complexity around named addresses as it prevents shadowing issues around module
+members (specifically constants) and local variables.
+
+In the old syntax, all address values began with `0x` and this hex prefix could not be used for
+integer values. In the new syntax, all address values are prefixed with `@`. Following the `@`
+operator any valid address can be used. For example:
+
+```move
+let _: u8 = 0x1u8;
+let _: u64 = 0x42u64;
+let _: u128 = 0x42u128;
+let a1: address = @Std;
+let a2: address = @66;
+let a3: address = @0x42;
+```
+
+In a way, you can think of `@` as an operator that takes an address from being a namespace item to
+an expression item.
+
+### Declaring Named Addresses
+
+In this new system, anywhere numerical address literals are used, a name can be used instead. New
+address names can be declared as:
+
+```move
+address MyAddr;
+```
+
+Using an address block will also declare the name:
+
+```move
+address MyAddr {
+module M {
+    ...
+}
+}
+```
+
+However, using the `address::module` syntax will not declare the name. It must be declared before
+used:
+
+```move
+address MyAddr;
+module MyAddr::M {
+    ...
+}
+```
+
+The declared address can then be used in both module access and or as expression values (with the
+new `@` syntax)
+
+```move
+script {
+    fun example() {
+        MyAddr::M::foo(@MyAddr);
+    }
+}
+```
+
+An address can be declared any number of times throughout a program.
+
+```move
+// file1.move
+address MyAddr;
+
+module MyAddr::M {
+    ...
+}
+```
+
+```move
+// file2.move
+address MyAddr {
+module N {
+    ...
+}
+}
+```
+
+### Assigning Named Addresses
+
+The address can be assigned a value with `= <number value>`:
+
+```move
+address MyAddr = 0xC0FFEE;
+```
+
+Or in an address block
+
+```move
+address MyAddr = 0xC0FFEE {
+module M {
+   ...
+}
+}
+```
+
+The address cannot be assigned using the `address::module` syntax. It must be assigned elsewhere:
+
+```move
+address MyAddr = 0xCOFFEE;
+module MyAddr::M {
+    ...
+}
+```
+
+An address can be assigned any number of times in a program, but it can only be given _one_ value.
+The following would be fine, since the address `MyAddr` is given the same value in both assignmets:
+
+```move
+address MyAddr = 0xC0FFEE;
+address MyAddr = 12648430; // decimal representation of 0xC0FFEE
+```
+
+Assigning a two different values will result in an error:
+
+```move
+address MyAddr = 0xCOFFEE;
+address MyAddr = 0xDEADBEEF; // ERROR!
+```
+
+### Opaqueness
+
+These assignments, and the name system as whole, only exist at the source language level. Names will
+be fully substituted for their value at the byte code level. So the example from before would be
+equivalent to
+
+```move
+script {
+    fun example() {
+        0xC0FFEE::M::foo(@0xC0FFEE);
+    }
+}
+```
+
+But at the source language level, the two are not interchangeable. If we had the declaration:
```

Comment:
Alluded to this in my question in the intro, but I think this point is perhaps surprising and should be made as early as possible.